lib/model/folder_sendrecv.go

```diff
@@ -1558,7 +1558,37 @@ func (f *sendReceiveFolder) performFinish(file, curFile protocol.FileInfo, hasCu
 			return err
 		}
 	} else if !fs.IsNotExist(err) {
-		return err
+		// If there's a case conflict, check if the conflicting files
+		// have the same content and "merge" (choose one, delete the other).
+		var caseErr *fs.ErrCaseConflict
+		if !errors.As(err, &caseErr) {
+			return err
+		}
+		// We can't deal with case conflict in parents (would have to check
+		// equality of the entire filesystem subtree).
+		if filepath.Dir(file.Name) != filepath.Dir(caseErr.Real) {
+			return err
+		}
+		// Check that the case conflict is up-to-date in db
+		stat, statErr := f.fs.Lstat(caseErr.Real)
+		if statErr != nil {
+			return statErr
+		}
+		other, ok := snap.Get(protocol.LocalDeviceID, caseErr.Real)
+		if scanErr := f.scanIfItemChanged(other.Name, stat, other, ok, scanChan); scanErr != nil {
+			return scanErr
+		}
+		if other.FileType() != file.FileType() || !other.BlocksEqual(file) {
+			return err
+		}
+		// Arbitrary but deterministic way of choosing one of the two names.
+		if other.Name > file.Name {
+			file.Name = other.Name
+		}
+		if err := f.inWritableDir(f.fs.Remove, other.Name); err != nil {
+			return err
+		}
+		scanChan <- other.Name // Scan the deletion
```

Comment:
I don't fully understand what happens here, can you explain? We decide to retain either the old or new name, potentially replacing the name of the thing we are syncing. We remove the file under old name, and queue a scan. We then continue syncing, potentially putting the file into the old name (if we chose that) and updating the database under that name.

It seems to me that this is essentially a no-op, in that we won't get a database update for the thing we were syncing (we changed the name) and we replaced the file that was there with an identical one under the same name, and we're back to square one?

Reply:
You are perfectly right, changing the name of the fileinfo does nothing respectively the wrong thing. I corrected it, but thinking more about it there's still unresolved problems that let me doubt whether this is a good idea at all, see https://github.com/syncthing/syncthing/pull/7014#issuecomment-701277474